### PR TITLE
Feature/fga/voip dial pad

### PR DESCRIPTION
**Pull request by ganfra**
This PR introduces: 
- DialPad for starting a call (with PSTN support)
- DialPad in call (with DTMF support)
-

### CODE REVIEW

**Review by dbkr — APPROVED** [review]

**Comment by ganfra** [comment]
Converted to draft waiting for last design inputs

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultSession.kt [hunk 1]
```diff
@@ -113,6 +114,7 @@ internal class DefaultSession @Inject constructor(
         private val accountService: Lazy<AccountService>,
         private val defaultIdentityService: DefaultIdentityService,
         private val integrationManagerService: IntegrationManagerService,
+        private val thirdPartyService: ThirdPartyService,
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultSession.kt:117]: Is there a reason to not put it Lazy  and use a getter to namespace the functionnalityN

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/thirdparty/ThirdPartyService.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.matrix.android.sdk.api.session.thirdparty
+
+import org.matrix.android.sdk.api.session.room.model.thirdparty.ThirdPartyProtocol
+import org.matrix.android.sdk.api.session.thirdparty.model.ThirdPartyUser
+
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/thirdparty/ThirdPartyService.kt:21]: Could add  https://matrix.org/docs/spec/client_server/r0.4.0.html#get-matrix-client-r0-thirdparty-protocols in doc

File: vector/build.gradle [hunk 3]
```diff
@@ -447,6 +447,8 @@ dependencies {
     implementation 'com.vanniktech:emoji-material:0.7.0'
     implementation 'com.vanniktech:emoji-google:0.7.0'
 
+    implementation 'im.dlg:android-dialer:1.2.5'
```
Comment (BillCarsonFr) [inline vector/build.gradle:450]: Is there an entry for it in open_source_licences.html?

File: vector/src/main/java/im/vector/app/features/call/CallAudioManager.kt [hunk 4]
```diff
@@ -100,8 +99,6 @@ class CallAudioManager(
         val audioManager = audioManager ?: return
         savedIsSpeakerPhoneOn = audioManager.isSpeakerphoneOn
         savedIsMicrophoneMute = audioManager.isMicrophoneMute
-        savedAudioMode = audioManager.mode
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/CallAudioManager.kt:103]: We don't have to restore the mode anymore? If i was listening to music on speaker before the call what happens after?

File: vector/src/main/java/im/vector/app/features/call/VectorCallActivity.kt [hunk 5]
```diff
@@ -114,7 +121,9 @@ class VectorCallActivity : VectorBaseActivity<ActivityCallBinding>(), CallContro
         if (intent.getStringExtra(EXTRA_MODE) == INCOMING_RINGING) {
             turnScreenOnAndKeyguardOff()
         }
-
+        if (savedInstanceState != null) {
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/VectorCallActivity.kt:124]: do that in the fragment onAttached?

File: vector/src/main/java/im/vector/app/features/call/VectorCallActivity.kt [hunk 6]
```diff
@@ -320,6 +328,11 @@ class VectorCallActivity : VectorBaseActivity<ActivityCallBinding>(), CallContro
             is VectorCallViewEvents.ConnectionTimeout -> {
                 onErrorTimoutConnect(event.turn)
             }
+            is VectorCallViewEvents.ShowDialPad -> {
+                CallDialPadBottomSheet.newInstance(false).apply {
+                    callback = dialPadCallback
+                }.show(supportFragmentManager, FRAGMENT_DIAL_PAD_TAG)
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/VectorCallActivity.kt:334]: no need to check if already there using tag?

File: vector/src/main/java/im/vector/app/features/call/VectorCallViewActions.kt [hunk 7]
```diff
@@ -25,7 +25,10 @@ sealed class VectorCallViewActions : VectorViewModelAction {
     object ToggleMute : VectorCallViewActions()
     object ToggleVideo : VectorCallViewActions()
     object ToggleHoldResume: VectorCallViewActions()
+    object OpenDialPad: VectorCallViewActions()
     data class ChangeAudioDevice(val device: CallAudioManager.SoundDevice) : VectorCallViewActions()
+    data class SendDtmfDigit(val digit: String) : VectorCallViewActions()
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/VectorCallViewActions.kt:30]: what's dtfm?

File: vector/src/main/java/im/vector/app/features/call/dialpad/CallDialPadBottomSheet.kt [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        super.onDestroyView()
+    }
+
+    private fun setCallbackToFragment(callback: DialPadFragment.Callback?) {
+        if (!isAdded) return
+        val dialPadFragment = childFragmentManager.findFragmentById(R.id.callDialPadFragmentContainer) as? DialPadFragment
+        dialPadFragment?.callback = DialPadFragmentCallbackWrapper(callback)
+    }
+
+    private inner class DialPadFragmentCallbackWrapper(val callback: DialPadFragment.Callback?): DialPadFragment.Callback {
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/dialpad/CallDialPadBottomSheet.kt:89]: why do we need a wrapper?

File: vector/src/main/java/im/vector/app/features/call/dialpad/DialPadFragment.kt [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+        digits?.isCursorVisible = cursorVisible
+        digits?.setTextColor(ThemeUtils.getColor(requireContext(), im.vector.app.R.attr.riotx_text_primary))
+        dialpadView.findViewById<View>(R.id.zero).setOnClickListener { append('0') }
+        if (enablePlus) {
+            dialpadView.findViewById<View>(R.id.zero).setOnLongClickListener {
+                append('+')
+                true
+            }
+        }
+        dialpadView.findViewById<View>(R.id.one).setOnClickListener { append('1') }
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/dialpad/DialPadFragment.kt:68]: we don't debounce here?

File: vector/src/main/java/im/vector/app/features/call/transfer/CallTransferActivity.kt [hunk 10]
```diff
@@ -86,76 +78,43 @@ class CallTransferActivity : VectorBaseActivity<ActivityCallTransferBinding>(),
...
-                            singleSelection = true,
-                            showInviteActions = false,
-                            showToolbar = false
-                    ),
-                    USER_LIST_FRAGMENT_TAG
-            )
-        }
+
         callTransferViewModel.observeViewEvents {
             when (it) {
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/transfer/CallTransferActivity.kt:83]: format ->

File: vector/src/main/java/im/vector/app/features/call/webrtc/WebRtcCallManager.kt [hunk 11]
```diff
@@ -88,11 +104,25 @@ class WebRtcCallManager @Inject constructor(
...
 
     private val rootEglBase by lazy { EglUtils.rootEglBase }
 
     private var isInBackground: Boolean = true
 
+    init {
+        GlobalScope.launch {
+            supportedPSTNProtocol = currentSession?.getSupportedPSTN(3)
+            if (supportedPSTNProtocol != null) {
+                pstnSupportListeners.forEach { it.onPSTNSupportUpdated() }
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/webrtc/WebRtcCallManager.kt:121]: would be safer to catch callbacks failures

File: vector/src/main/java/im/vector/app/features/matrixto/MatrixToBottomSheetViewModel.kt [hunk 12]
```diff
@@ -125,42 +124,23 @@ class MatrixToBottomSheetViewModel @AssistedInject constructor(
...
 
     private fun handleStartChatting(action: MatrixToAction.StartChattingWithUser) {
-        val mxId = action.matrixItem.id
-        val existing = session.getExistingDirectRoomWithUser(mxId)
-        if (existing != null) {
-            // navigate to this room
-            _viewEvents.post(MatrixToViewEvents.NavigateToRoom(existing))
-        } else {
+        viewModelScope.launch {
             setState {
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/matrixto/MatrixToBottomSheetViewModel.kt:128]: Set loading state before launching in coroutine scope

File: vector/src/main/java/im/vector/app/features/roomdirectory/picker/RoomDirectoryPickerViewModel.kt [hunk 13]
```diff
@@ -52,19 +52,18 @@ class RoomDirectoryPickerViewModel @AssistedInject constructor(@Assisted initial
...
 
     private fun load() {
-        session.getThirdPartyProtocol(object : MatrixCallback<Map<String, ThirdPartyProtocol>> {
-            override fun onSuccess(data: Map<String, ThirdPartyProtocol>) {
+        viewModelScope.launch {
+            try {
+                val thirdPartyProtocols = session.getThirdPartyProtocols()
                 setState {
-                    copy(asyncThirdPartyRequest = Success(data))
+                    copy(asyncThirdPartyRequest = Success(thirdPartyProtocols))
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/roomdirectory/picker/RoomDirectoryPickerViewModel.kt:59]: No Loading() state?

File: vector/src/main/res/drawable/ic_call_dial_pad.xml [hunk 14]
```diff
@@ -0,0 +1,9 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="16dp"
+    android:height="22dp"
+    android:viewportWidth="16"
+    android:viewportHeight="22">
+  <path
+      android:pathData="M8,18C6.9,18 6,18.9 6,20C6,21.1 6.9,22 8,22C9.1,22 10,21.1 10,20C10,18.9 9.1,18 8,18ZM2,0C0.9,0 0,0.9 0,2C0,3.1 0.9,4 2,4C3.1,4 4,3.1 4,2C4,0.9 3.1,0 2,0ZM2,6C0.9,6 0,6.9 0,8C0,9.1 0.9,10 2,10C3.1,10 4,9.1 4,8C4,6.9 3.1,6 2,6ZM2,12C0.9,12 0,12.9 0,14C0,15.1 0.9,16 2,16C3.1,16 4,15.1 4,14C4,12.9 3.1,12 2,12ZM14,4C15.1,4 16,3.1 16,2C16,0.9 15.1,0 14,0C12.9,0 12,0.9 12,2C12,3.1 12.9,4 14,4ZM8,12C6.9,12 6,12.9 6,14C6,15.1 6.9,16 8,16C9.1,16 10,15.1 10,14C10,12.9 9.1,12 8,12ZM14,12C12.9,12 12,12.9 12,14C12,15.1 12.9,16 14,16C15.1,16 16,15.1 16,14C16,12.9 15.1,12 14,12ZM14,6C12.9,6 12,6.9 12,8C12,9.1 12.9,10 14,10C15.1,10 16,9.1 16,8C16,6.9 15.1,6 14,6ZM8,6C6.9,6 6,6.9 6,8C6,9.1 6.9,10 8,10C9.1,10 10,9.1 10,8C10,6.9 9.1,6 8,6ZM8,0C6.9,0 6,0.9 6,2C6,3.1 6.9,4 8,4C9.1,4 10,3.1 10,2C10,0.9 9.1,0 8,0Z"
```
Comment (BillCarsonFr) [inline vector/src/main/res/drawable/ic_call_dial_pad.xml:7]: 18.7 ? 
:trollface:

**Review by BillCarsonFr — COMMENTED** [review]
Just a few comments

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultSession.kt (hunk 1 above)
Comment (ganfra) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/DefaultSession.kt:117]: No reason, I switch to your proposal.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/thirdparty/ThirdPartyService.kt (hunk 2 above)
Comment (ganfra) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/thirdparty/ThirdPartyService.kt:21]: Done

File: vector/build.gradle (hunk 3 above)
Comment (ganfra) [reply vector/build.gradle:450]: Nice catch. But it'll probably be removed by a custom design. Still, I add it.

File: vector/src/main/java/im/vector/app/features/call/CallAudioManager.kt (hunk 4 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/call/CallAudioManager.kt:103]: During my tests it was ok. I guess the app will get focus again

File: vector/src/main/java/im/vector/app/features/call/VectorCallActivity.kt (hunk 5 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/call/VectorCallActivity.kt:124]: What does it change, I still need to access the callback

File: vector/src/main/java/im/vector/app/features/call/VectorCallActivity.kt (hunk 6 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/call/VectorCallActivity.kt:334]: It's not really possible to be in this case so I guess its ok

File: vector/src/main/java/im/vector/app/features/call/VectorCallViewActions.kt (hunk 7 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/call/VectorCallViewActions.kt:30]: https://fr.wikipedia.org/wiki/Code_DTMF#:~:text=Un%20code%20DTMF%20(dual%2Dtone,(sauf%20voix%20sur%20IP). :D

File: vector/src/main/java/im/vector/app/features/call/dialpad/CallDialPadBottomSheet.kt (hunk 8 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/call/dialpad/CallDialPadBottomSheet.kt:89]: It's just to be sure we call dismiss

File: vector/src/main/java/im/vector/app/features/call/dialpad/DialPadFragment.kt (hunk 9 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/call/dialpad/DialPadFragment.kt:68]: DOn't think it's necessary...

File: vector/src/main/java/im/vector/app/features/call/webrtc/WebRtcCallManager.kt (hunk 11 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/call/webrtc/WebRtcCallManager.kt:121]: Yup, thanks

File: vector/src/main/res/drawable/ic_call_dial_pad.xml (hunk 14 above)
Comment (ganfra) [reply vector/src/main/res/drawable/ic_call_dial_pad.xml:7]: AH

File: vector/src/main/java/im/vector/app/features/matrixto/MatrixToBottomSheetViewModel.kt (hunk 12 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/matrixto/MatrixToBottomSheetViewModel.kt:128]: why?

**Pull request merged**